src/node/historical_queries.h

```diff
@@ -18,36 +22,74 @@ namespace ccf::historical
   class StateCache : public AbstractStateCache
   {
   protected:
-    kv::Store& source_store;
+    ccf::NetworkState& network;
     ringbuffer::WriterPtr to_host;
 
+    std::shared_ptr<ccf::LedgerSecrets> historical_ledger_secrets;
+    std::shared_ptr<ccf::NodeEncryptor> historical_encryptor;
+
     enum class RequestStage
     {
+      RecoveringLedgerSecret,
       Fetching,
       Untrusted,
       Trusted,
     };
 
     using LedgerEntry = std::vector<uint8_t>;
 
+    struct LedgerSecretRecoveryInfo
+    {
+      consensus::Index target_idx = 0;
+      LedgerSecretPtr last_ledger_secret;
+
+      LedgerSecretRecoveryInfo(
+        consensus::Index target_idx_, LedgerSecretPtr last_ledger_secret_) :
+        target_idx(target_idx_),
+        last_ledger_secret(last_ledger_secret_)
+      {}
+    };
+
     struct Request
     {
       RequestStage current_stage = RequestStage::Fetching;
       crypto::Sha256Hash entry_hash = {};
       StorePtr store = nullptr;
+
+      // Only set when recovering ledger secrets
+      std::unique_ptr<LedgerSecretRecoveryInfo> ledger_secret_recovery_info =
+        nullptr;
     };
 
     // These constitute a simple LRU, where only user queries will refresh an
     // entry's priority
     static constexpr size_t MAX_ACTIVE_REQUESTS = 10;
-    std::map<consensus::Index, Request> requests;
+
+    using Requests = std::map<consensus::Index, Request>;
+    Requests requests;
     std::list<consensus::Index> recent_requests;
 
     // To trust an index, we currently need to fetch a sequence of entries
     // around it - these aren't user requests, so we don't store them, but we do
     // need to distinguish things-we-asked-for from junk-from-the-host
     std::set<consensus::Index> pending_fetches;
 
+    ccf::VersionedLedgerSecret get_first_known_ledger_secret()
+    {
+      auto tx = network.tables->create_read_only_tx();
```

Comment:
I think we should avoid creating this tx here, and instead have `get_latest`/`get_first` variants that don't take a `Tx` and don't have a transactional dependency, iff doing this non-transactionally is really safe.

At the very least, we should create 2 different transactions for passing to the different stores: The assert means this `Tx` holds views over 2 different tables of the same name on different stores (in Debug only!). This is completely unexpected behaviour, and its probably doing something wrong under-the-hood.

Reply:
The transaction created here is read-only and never commits so it is only created to match the current interface. Unless I'm missing something, there's only one store involved here though (the main one, i.e. `network.tables`) so I think it is preferable to keep the current interface as it is?

Reply:
Yeah you're right, sorry I was way off here. I thought `LedgerSecrets` had a reference to the `Store` and was getting secrets from that store's tables, but of course this doesn't happen - the tables are only referred to by name, they're not per-Store.

I'm tempted to say that `get_first` doesn't need a `Tx` though, and then this `tx` is still unnecessary? `get_latest` is a cache/interpretation of the current value, so needs a transactional dependency to make sure its still 'current'. But `first` is an independent cache unrelated to the transaction - if `first` changes, it doesn't need to invalidate any in-flight transactions?

Reply:
Yep. I've now removed the `tx` argument to `get_first()`.